vm/src/builtins/set.rs

```diff
@@ -649,49 +652,103 @@ impl PySet {
 
     #[pymethod]
     fn intersection_update(
-        &self,
-        others: PosArgs<ArgIterable>,
+        zelf: PyRef<Self>,
+        others: PosArgs<PyObjectRef>,
         vm: &VirtualMachine,
     ) -> PyResult<()> {
-        self.inner.intersection_update(others.into_iter(), vm)?;
+        let arguments = others.into_vec();
+        if arguments.len() == 1 {
+            if let Some(iterable) = arguments.first() {
+                if zelf.is(iterable) {
+                    return Ok(());
+                }
+            }
+        } else if arguments.is_empty() {
+            return Ok(());
+        }
+        zelf.inner.intersection_update(arguments.into_iter(), vm)?;
         Ok(())
     }
 
     #[pymethod(magic)]
-    fn iand(zelf: PyRef<Self>, set: AnySet, vm: &VirtualMachine) -> PyResult<PyRef<Self>> {
-        zelf.inner
-            .intersection_update(std::iter::once(set.into_iterable(vm)?), vm)?;
+    fn iand(zelf: PyRef<Self>, set: PyObjectRef, vm: &VirtualMachine) -> PyResult<PyRef<Self>> {
+        if !set.class().is(vm.ctx.types.set_type) {
+            return Err(vm.new_type_error("Type error".to_owned()));
```

Comment:
Should be `unsupported operand type(s) for +=: 'type1' and 'type2'"`, similarly for the rest of the type errors.